src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php

```diff
@@ -274,12 +288,54 @@ private function getCategoriesInformation(Product $product): CategoriesInformati
 
         $categoriesInformation = [];
         foreach ($categoryNames as $categoryId => $localizedNames) {
-            $categoriesInformation[] = new CategoryInformation($categoryId, $localizedNames);
+            $categoryName = $categoryNames[$categoryId][$this->contextLangId];
+            $displayName = $this->buildDisplayName(
+                new CategoryId($categoryId),
+                $languageId,
+                $categoryName,
+                $duplicateNames
+            );
+            $categoriesInformation[] = new CategoryInformation(
+                $categoryId,
+                $categoryName,
+                $displayName
+            );
         }
 
         return new CategoriesInformation($categoriesInformation, $defaultCategoryId);
     }
 
+    /**
+     *  If there are multiple categories with identical names, we want to be able to tell them apart,
+     *  so we use breadcrumb path instead of category name.
+     *  However, whole breadcrumb path would probably be too long, therefore not UX friendly.
+     *  Calculating "optimal" breadcrumb length seems too complex compared to the value it could bring.
+     *  So, we show one parent name and category name, as it is simple and should cover most cases.
+     *
+     * e.g. "Clothes > Women"
+     *
+     * @param CategoryId $categoryId
+     * @param LanguageId $languageId
+     * @param string $categoryName
+     * @param string[] $duplicateNames
+     *
+     * @return string
+     */
+    private function buildDisplayName(
```

Comment:
I think this technical feature could be handled in a separate class because you do almost the same thing in the tree handler. I imagine a service `CategoryDisplayNameBuilder` which would internally fetch the `duplicateNames` by using `Category::getDuplicateNames` internally. And a single public method:

```php
public function buidDisplayName(string $categoryName, bool $refreshDuplicates = false): string
```

By default, the duplicated names is fetched once and stored in a class field, but the `$refreshDuplicates` can force fetching it. I think about this for the test issues, because we need some kind of cache that can be cleared easily. Maybe you should use a cache adapter or the legacy `Cache` class but I'm less fan of the legacy approach.